### PR TITLE
[WIP]Fast Finality: reward distribution and slash parts

### PR DESCRIPTION
**Pull request by Loverush**
### Description

add a description of your changes here...

### Rationale

tell us why we need these changes...

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### CODE REVIEW

File: consensus/parlia/parlia.go [hunk 1]
```diff
@@ -1151,6 +1156,62 @@ func (p *Parlia) distributeToValidator(amount *big.Int, validator common.Address
 	return p.applyTransaction(msg, state, header, chain, txs, receipts, receivedTxs, usedGas, mining)
 }
 
+// distribute finality reward
+func (p *Parlia) distributeFinalityReward(chain consensus.ChainHeaderReader, currentHeight uint64,
+	state *state.StateDB, header *types.Header, chainContext core.ChainContext, txs *[]*types.Transaction,
+	receipts *[]*types.Receipt, receivedTxs *[]*types.Transaction, usedGas *uint64, mining bool) error {
+	accumulatedWeights := make(map[common.Address]uint64)
+	for i := currentHeight - p.config.Epoch + 1; i <= currentHeight; i++ {
+		header := chain.GetHeaderByNumber(i)
```
Comment (realuncle) [inline consensus/parlia/parlia.go:1165]: Please handle if `header` is nil.

File: consensus/parlia/parlia.go [hunk 2]
```diff
@@ -1151,6 +1156,62 @@ func (p *Parlia) distributeToValidator(amount *big.Int, validator common.Address
...
+		assembledMessage := header.Finality
+		if assembledMessage == nil {
+			continue
+		}
+		finalizedBlock := chain.GetHeaderByHash(assembledMessage.Data.BlockHash)
```
Comment (realuncle) [inline consensus/parlia/parlia.go:1170]: Please rename `finalizedBlock ` ==> `justifiedBlock`

File: core/vm/contracts.go [hunk 3]
```diff
@@ -21,14 +21,19 @@ import (
...
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/crypto/blake2b"
 	"github.com/ethereum/go-ethereum/crypto/bls12381"
 	"github.com/ethereum/go-ethereum/crypto/bn256"
 	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/prysmaticlabs/prysm/crypto/bls"
```
Comment (calmbeing) [inline core/vm/contracts.go:35]: Please put the third part package in the middle group

File: consensus/parlia/parlia.go [hunk 4]
```diff
@@ -682,7 +682,7 @@ func (p *Parlia) PrepareVoteAttestation(chain consensus.ChainHeaderReader, heade
 		return nil
 	}
 
-	var attestation types.VoteAttestation
+	var attestation types.VoteEnvelope
```
Comment (calmbeing) [inline consensus/parlia/parlia.go:685]: will `*types.Envelope` be better?

File: consensus/parlia/parlia.go [hunk 5]
```diff
@@ -776,6 +776,73 @@ func (p *Parlia) verifyValidators(header *types.Header) error {
...
+			return err
+		}
+		validators := snap.validators()
+		for j := 0; j < 64; j++ {
+			if ((uint64(voteAttestation.VoteAddressSet) >> j) & 1) == 1 {
+				accumulatedWeights[validators[j]] += rewardCoef
+			}
+		}
+	}
+	validators := make([]common.Address, len(accumulatedWeights))
```
Comment (calmbeing) [inline consensus/parlia/parlia.go:821]: What about using make([]common.Address, 0, len(accumulatedWeights))?

File: consensus/parlia/parlia.go [hunk 6]
```diff
@@ -776,6 +776,73 @@ func (p *Parlia) verifyValidators(header *types.Header) error {
...
+	weights := make([]uint64, len(accumulatedWeights))
```
Comment (calmbeing) [inline consensus/parlia/parlia.go:822]: Same with above

File: consensus/parlia/parlia.go (hunk 4 above)
Comment (Loverush) [reply consensus/parlia/parlia.go:685]: Fixed

File: consensus/parlia/parlia.go (hunk 5 above)
Comment (Loverush) [reply consensus/parlia/parlia.go:821]: Fixed

File: consensus/parlia/parlia.go (hunk 6 above)
Comment (Loverush) [reply consensus/parlia/parlia.go]: Fixed

**Pull request merged**